### PR TITLE
pe: Add checksum field, function to calculate the checksum

### PR DESCRIPTION
**Pull request by hillu**

### CODE REVIEW

File: libyara/modules/pe.c [hunk 1]
```diff
@@ -1849,6 +1853,52 @@ define_function(rich_toolid_version)
...
+
+define_function(calculate_checksum)
+{
+  YR_OBJECT* module = module();
+
+  PE* pe = (PE*) module->data;
+  if (pe == NULL)
+    return_integer(UNDEFINED);
+
+  uint64_t csum = 0;
```
Comment (plusvic) [inline libyara/modules/pe.c]: In C99 variables can be declared anywhere within a block, but in C89 they must be at the beginning of the block. This has been a source of problems with Visual Studio 2010.

File: libyara/modules/pe.c [hunk 2]
```diff
@@ -1849,6 +1853,52 @@ define_function(rich_toolid_version)
...
+  int csum_offset = ((uint8_t*)&(pe->header->OptionalHeader) +
+		     offsetof(IMAGE_OPTIONAL_HEADER32, CheckSum)) - pe->data;
+  for (int i=0; i <= pe->data_size / 4; i++)
+  {
+    // Treat the CheckSum field as 0 -- the offset is the same for
+    // PE32 and PE64.
+    if (4*i == csum_offset)
```
Comment (plusvic) [inline libyara/modules/pe.c]: Spaces around the *

File: libyara/modules/pe.c [hunk 3]
```diff
@@ -1849,6 +1853,52 @@ define_function(rich_toolid_version)
...
+  YR_OBJECT* module = module();
+
+  PE* pe = (PE*) module->data;
+  if (pe == NULL)
+    return_integer(UNDEFINED);
+
+  uint64_t csum = 0;
+  int csum_offset = ((uint8_t*)&(pe->header->OptionalHeader) +
+		     offsetof(IMAGE_OPTIONAL_HEADER32, CheckSum)) - pe->data;
+  for (int i=0; i <= pe->data_size / 4; i++)
```
Comment (plusvic) [inline libyara/modules/pe.c]: Spaces around the =

File: libyara/modules/pe.c [hunk 4]
```diff
@@ -1849,6 +1853,52 @@ define_function(rich_toolid_version)
...
+  {
+    // Treat the CheckSum field as 0 -- the offset is the same for
+    // PE32 and PE64.
+    if (4*i == csum_offset)
+      continue;
+    if (4*i+4 < pe->data_size)
```
Comment (plusvic) [inline libyara/modules/pe.c]: Spaces around * and +

File: libyara/modules/pe.c [hunk 5]
```diff
@@ -1849,6 +1853,52 @@ define_function(rich_toolid_version)
...
+    {
+      csum += ((uint64_t)pe->data[4*i] +
```
Comment (plusvic) [inline libyara/modules/pe.c]: Space after the type cast. e.g: (uint64_t) xxxx

File: libyara/modules/pe.c [hunk 6]
```diff
@@ -1849,6 +1853,52 @@ define_function(rich_toolid_version)
...
+	       ((uint64_t)pe->data[4*i+3] << 24));
+    }
+    else
+    {
+      switch (pe->data_size % 4)
+      {
+        case 3: csum += (uint64_t)pe->data[4*i+2] << 16;
+        case 2: csum += (uint64_t)pe->data[4*i+1] << 8;
+        case 1: csum += (uint64_t)pe->data[4*i];
+      }
```
Comment (plusvic) [inline libyara/modules/pe.c]: Clever use of the switch fall through, but I think a for loop would make the intention more evident.

**Comment by hillu** [comment]
I am not sure what to make of "clever" there. ;-)

Anyway, I have updated my patch.

**Comment by wxsBSD** [comment]
Thanks for writing this! Gloriana asked for something similar yesterday, but I wasn't sure you were going to get to it so fast!

**Pull request merged**